src/test/kotlin/com/atlassian/performance/tools/jiraactions/api/JqlMemoryTest.kt

```diff
@@ -0,0 +1,73 @@
+package com.atlassian.performance.tools.jiraactions.api
+
+import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory
+import com.atlassian.performance.tools.jiraactions.api.memories.Project
+import com.atlassian.performance.tools.jiraactions.api.memories.adaptive.AdaptiveJqlMemory
+import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
+import com.atlassian.performance.tools.jiraactions.jql.BuiltInJQL
+import org.assertj.core.api.Assertions
+import org.junit.Test
+import org.mockito.Mockito
+import org.mockito.stubbing.Answer
+import org.openqa.selenium.WebDriver
+
+class JqlMemoryTest {
+
+    @Test
+    fun allBuiltInQueryTypesAvailable() {
+        val memory: JqlMemory = AdaptiveJqlMemory(SeededRandom());
+        val webdriverMock = Mockito.mock(WebDriver::class.java)
+        val issuePageSpy = Mockito.spy(IssuePage(webdriverMock))
```

Comment:
This is unfortunately necessary if we want to mock `IssuePage`.
Ideally, we would turn `IssuePage` into an interface and use a mock implementation here, avoiding the huge runtime, dependency, readability and diagnosability cost of Mockito.

Changing a class into an interface is binary-incompatible (`invokeinterface` bytecode instruction). Well be able to clean this up only in a major release.

Reply:
I do not see reason to write more code or do refactoring when my objective different. I see mockito is widely used in atlassian/spartez projets. Because of mockito this test case is clean and neat. 
I dont't get why we would be able to clean this in major release - this is the same module.

Reply:
I think runtime and dependency costs doesn't need explanation. I want to add something about readability and diagnosability:

- A proper mock could be reused across multiple tests. We have just two tests here, but we need to repeat the blocks of code that creates the mock.
- It can introduce bugs in the future. Without Mockito, when changing mocked objects, you need to adjust your mocks. You need to make a conscious decision, how the mock should behave.  Mockito can transparently adapt for most cases. In the end, you may end up testing something else than intended (we hade examples of the cases like that internally). 
- Like every reflection-based tool, it may fool IDE. Searching for usages of a method may not work well (I'm not sure if it's still an issue in IJ 2019, but it's an open-source project and developers can use any IDE).
- If you put complex logic into your mock, then it's hard to debug (reflection again). It's ok if you return values, but by introducing the library on a classpath, you encourage developers to use it for every case.
- We have multiple reasons to write tests. To protect against regressions. To document how to use an API. To reproduce the issue. Another reason is to verify if the API is useable. Mockito makes mocking unmockable objects easy. It means we'll stop to see problematic places and we'll never improve it. Problems with testing are very similar to issues of the code and reuse of it. We should treat our tests as the first user of an API.

Reply:
> * A proper mock could be reused across multiple tests. We have just two tests here, but we need to repeat the blocks of code that creates the mock.

I used to create mock to cover some specific code path and i think it is right.
You're right about redundant mock preparation code.

>* It can introduce bugs in the future. Without Mockito, when changing mocked objects, you need to adjust your mocks. You need to make a conscious decision, how the mock should behave. Mockito can transparently adapt for most cases. In the end, you may end up testing something else than intended (we had examples of the cases like that internally).

Mockito requires using explicit API calls to prepare mocked instance so it is cean how API is used and mocked. Involving mockito or not has nothing to do with test scope/goal/intention.

> * Like every reflection-based tool, it may fool IDE. Searching for usages of a method may not work well (I'm not sure if it's still an issue in IJ 2019, but it's an open-source project and developers can use any IDE).

There are explicit API calls

> * If you put complex logic into your mock, then it's hard to debug (reflection again). It's ok if you return values, but by introducing the library on a classpath, you encourage developers to use it for every case.

There is no complex logic in here

> * We have multiple reasons to write tests. To protect against regressions. To document how to use an API. To reproduce the issue. Another reason is to verify if the API is useable. Mockito makes mocking unmockable objects easy. It means we'll stop to see problematic places and we'll never improve it. Problems with testing are very similar to issues of the code and reuse of it. We should treat our tests as the first user of an API.

This is separate concern (unmockable objects and its improvement) and mockito does not caused the problem and does not solves it (in terms of design). When you wrote dozen lines of code without mockito and it is working for you, then this also causes problematic places are not visible.
Mockito helps to show how api can be used . Ido not see this is broken in area of test concern

Reply:
>I used to create mock to cover some specific code path and i think it is right.
You're right about redundant mock preparation code.

It's ok. It just violates the DRY principle. For some cases, redundancy can be better than a broken abstraction. I would not put much weight to that.

> Mockito requires using explicit API calls to prepare mocked instance so it is cean how API is used and mocked. Involving mockito or not has nothing to do with test scope/goal/intention.

This is an internal example of the problem I've mentioned https://hello.atlassian.net/wiki/spaces/~kcichy/blog/2018/10/12/321652230/Do+you+trust+your+mock#Mocks-go-wrong-over-time. It's clear how to use and mock objects. It'll work for most cases at the time when the test is written. But Mockito helps mocks to deteriorate unnoticeably.


> There is no complex logic in here


I'm not concerned with complex logic here. I'm concerned about giving a tool (on a classpath) that can be misused in the future.

> This is separate concern (unmockable objects and its improvement) and mockito does not caused the problem and does not solves it (in terms of design). When you wrote dozen lines of code without mockito and it is working for you, then this also causes problematic places are not visible.
Mockito helps to show how api can be used . Ido not see this is broken in area of test concern

 
Mockito makes dealing with problematic API easy. Writing dozens of lines of code is painfull, so it reminds about the problem. It's still easier to write the mocks than to fix the API, but the problem lays in a different place. The problem may be invisible (it's easy to mock with Mockito) for a new code. We may introduce a new broken API that's hard to mock because we have a tool with super-powers.

Reply:
@wyrzyk 
I do not see that reason for the problem with test mentioned in the blogpost of Kamil was mockito. If the test was written with manual mocks results would be the same.
Primary reason that the test didn't catch the problem is not implemented coverage.

Reply:
All test stuff is reverted. Not because I do agree with this review, but because I'm forced by time.

Reply:
@drspartez I think there's been a misunderstanding. I agreed that Mockito is necessary in this case and that we should go for it. Having a test with Mockito is better than not having a test at all.

I meant to inform that we wouldn't have to use Mockito if `IssuePage` was better designed. We can improve its design only in a major release. Then we should remove Mockito.

Being forced to resort to reflection hacks like Mockito proves how hard it is to customize `IssuePage`. Currently users have to settle on copy-pasting implementations.

TL;DR: Mockito is bad, but this time it was necessary.